src/cli.rs

```diff
@@ -553,13 +500,34 @@ impl Cli {
     }
 
     pub fn into_app() -> clap::App<'static> {
-        let mut app = <Self as clap::IntoApp>::into_app();
-        for &flag in NEGATION_FLAGS {
-            // `orig` and `flag` both need a static lifetime, so we
-            // build `orig` by trimming `flag` instead of building `flag`
-            // by extending `orig`
-            let orig = flag.strip_prefix("--no-").unwrap();
-            app = app.arg(
+        let app = <Self as clap::IntoApp>::into_app();
+
+        // Every option should have a --no- variant that makes it as if it was
+        // never passed.
+        // https://github.com/clap-rs/clap/issues/815
+        // https://github.com/httpie/httpie/blob/225dccb2186f14f871695b6c4e0bfbcdb2e3aa28/httpie/cli/argparser.py#L312
+        // Unlike HTTPie we apply the options in order, so the --no- variant
+        // has to follow the original to apply. You could have a chain of
+        // --x=y --no-x --x=z where the last one takes precedence.
+
+        let opts: Vec<_> = app.get_arguments().filter(|a| !a.is_positional()).collect();
+
+        // The strings in the `Arg`s need to live for 'static. That's a problem,
+        // because we also need to generate them right here.
+        // We could use Box::leak(), but this OnceCell maneuver keeps valgrind
+        // happy-ish.
+        // We assume that `get_arguments()` has a fixed iteration order.
+        static ARG_STORAGE: OnceCell<Vec<String>> = OnceCell::new();
+        let arg_storage = ARG_STORAGE.get_or_init(|| {
+            opts.iter()
+                .map(|opt| format!("--no-{}", opt.get_long().expect("long option")))
+                .collect()
+        });
```

Comment:
It would be great to mention this workaround in https://github.com/clap-rs/clap/issues/815